### ISSUE TITLE
Threading issue in waflib/extras/cpplint.py

### ISSUE THREAD

**Issue by danbar** [waf-project/waf#2174]
If I build and run Waf using
```
$ ./waf --tools=cpplint
$ ./waf configure build -j32
```
on a project generated by the project generator in the Waf source tree
```
$ ./utils/genbench.py /tmp/build 200 100 15 5
```
with the following wscript file
```python
#! /usr/bin/env python
# encoding: utf-8

VERSION = '0.0.2'
APPNAME = 'build_bench'
top  = '.'
out  = '.'

def options(opt):
	opt.load('cpplint')

def configure(conf):
	conf.load('cpplint')
	conf.env.CPPLINT_FILTERS = ','.join(('-')) # disable all filters
	conf.env.CPPLINT_BREAK = 1

def build(bld):
	for i in range(200):
		filez = ' '.join(['lib_%d/class_%d.h' % (i, j) for j in range(100)])
		bld(
			features = 'cpplint',
			source = filez,
			includes = '.', # include the top-level
		)
```
I can trigger a threading issue in `waflib/extras/cpplint.py`.
It means that a CppLint error is raised, although all CppLint errors are disabled in the wscript file.

The issue seems to be that the call `cpplint_tool._SetFilters(self.env.CPPLINT_FILTERS)` in `waflib/extras/cpplint.py` in line 189 is not thread safe.

**Comment by danbar** [comment]
Would it make sense to use the `multiprocessing` module for spawning a subprocess from your point of view?

**Comment by ita1024** [comment]
And what would be the benefits of adding `multiprocessing` as opposed to just running the process? Have you actually tried?

**Comment by danbar** [comment]
Yes, you are right. There is no benefit in using `multiprocessing`.
I have tried it out in my [branch](https://github.com/danbar/waf/tree/concurrency) using `Utils.subprocess.Popen`. Feedback is welcome.
